src/protocol/boolean/prefix_or.rs

```diff
@@ -0,0 +1,400 @@
+use crate::error::BoxError;
+use crate::ff::BinaryField;
+use crate::protocol::{attribution::IterStep, context::ProtocolContext, mul::SecureMul, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+use std::iter::{repeat, zip};
+
+/// This is an implementation of Prefix-Or on bitwise-shared numbers.
+///
+/// `PrefixOr` takes inputs `[a_1]_p,...,[a_l]_p where a1,...,a_l ∈ {0,1} ⊆ F_p`,
+/// and computes `[b_1]_p,...,[b_l]_p where b_i = ∨ a_j where j=1..i`.
+///
+/// In other words, `b_B <- PrefixOr(a_B)` will output `b_B` where b's i'th bit is
+/// 1 iff a's i'th bit is the first bit to have 1.
+///
+/// This is a sub-protocol used by Bitwise Less-Than protocol for an inequality check
+/// operation.
+///
+/// 5.2 Prefix-Or
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct PrefixOr<'a, B: BinaryField> {
+    input: &'a [Replicated<B>],
+}
+
+impl<'a, B: BinaryField> PrefixOr<'a, B> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a [Replicated<B>]) -> Self {
+        Self { input }
+    }
+
+    /// Securely computes `[a] | [b] where a, b ∈ {0, 1}`
+    /// OR can be computed as: `[a] ^ [b] ^ ([a] & [b])`
+    async fn bit_or(
+        a: Replicated<B>,
+        b: Replicated<B>,
+        ctx: ProtocolContext<'_, Replicated<B>, B>,
+        record_id: RecordId,
+    ) -> Result<Replicated<B>, BoxError> {
+        let a_and_b = ctx.multiply(record_id, a, b).await?;
+        Ok(a + b + a_and_b)
+    }
+
+    /// Securely computes `∨ [a_1],...[a_n]`
+    async fn block_or(
+        a: &[Replicated<B>],
+        k: usize,
+        ctx: ProtocolContext<'_, Replicated<B>, B>,
+        record_id: RecordId,
+    ) -> Result<Replicated<B>, BoxError> {
+        #[allow(clippy::cast_possible_truncation)]
+        let mut step = IterStep::new("bit", k as u32);
+        let mut v = Replicated::new(B::ZERO, B::ZERO);
+        for &bit in a {
+            v = Self::bit_or(v, bit, ctx.narrow(step.next()), record_id).await?;
```

Comment:
This does one extra round of communication - because you're actually invoking the secure multiply operation to multiply with a known zero value. Instead of initialising v to Replicated::default(), why not just initialise to the first element of a?